rdrf/rdrf/reports/generator.py

```diff
@@ -0,0 +1,707 @@
+        from explorer.models import Query
+
+        starting_models = [ContentType, Registry, Group, State, AddressType, NextOfKinRelationship,
+                           PatientAddress, Query, Section, ConsentSection, ConsentQuestion,ConsentValue,
+                           RegistryForm, CDEPermittedValue,
+                           CDEPermittedValueGroup, CommonDataElement, Patient]
+
+        if self.reporting_engine is self.default_engine:
+            raise Exception("reporting db = default!")
+
+        models = []
+
+        def exists(model):
+            return model.__name__ in [m.__name__ for m in models]
+
+        def add_model(model):
+            for m in self._get_related_models(model):
+                if m.__name__ == model.__name__:
+                    continue
+                add_model(m)
+            if not exists(model):
+                models.append(model)
+
+        for model in starting_models:
+            add_model(model)
+
+        def clone_model(model):
+            table_name = model._meta.db_table
+            self._mirror_table(
+                table_name, self.default_engine, self.reporting_engine)
+            logger.debug("mirrored table %s OK" % table_name)
+
+        def add_models(models):
+            bad = []
+            for model in models:
+                if model is not None:
+                    try:
+                        clone_model(model)
+                    except:
+                        bad.append(model)
+            return bad
+
+        finished = False
+        n = 1
+        while not finished and n < 100:
+            models = add_models(models)
+            finished = len(models) == 0
+            n += 1
+
+        if not finished:
+            raise Exception("Could not dump all demographic models: %s" % [
+                            m.__name__ for m in models])
+
+        logger.info("Dumped all demographic data OK")
+
+    def _get_related_models(self, model):
+        models = [model]
+        for field in model._meta.related_objects:
+            related_model = field.related_model
+            if related_model not in models:
+                models.append(related_model)
+            models.extend(self._get_related_models(related_model))
+
+        return models
+
+    def _mirror_table(self, table_name, source_engine, target_engine):
+        try:
+            self._drop_table(table_name)
+        except:
+            pass
+
+        source_meta = MetaData(bind=source_engine)
+        table = alc.Table(table_name, source_meta, autoload=True)
+        table.metadata.create_all(self.reporting_engine)
+        self._copy_table_data(source_engine, target_engine, table)
+
+    def create_tables(self):
+        if self.reporting_engine is not self.default_engine:
+            logger.info("CREATING DEMOGRAPHIC TABLES")
+            self._create_demographic_tables()
+
+        logger.info("CREATING CLINICAL TABLES")
+
+        for form_model in self.registry_model.forms:
```

Comment:
Extract `_create_clinical_tables` method to have methods working at the same level.  (ie. `_create_demographic_tables`, `_create_clinical_tables`, and `_extract_clinical_data`)